### DISCUSSION TITLE
Support of long/integer in ILP solver

### DISCUSSION THREAD

**Discussion by ahonorat** [optimatika/ojAlgo#425] [General, ojAlgo and related code]
Hello,

## Problem

Recently I've been facing a StackOverflowException when using ojAlgo version 44 to solve an ILP problem built with the ExpressionBasedModel class.

I guess this is due to an integer overflow, indeed the variable coefficients are large numbers close to the integer limit (2^31).
Moreover I had troubles setting upper limit to some variables to be minimized (if upper limit is above 2^31, the solution is clearly not optimal).  Maybe somewhere a loop/recusrion is bounded by an integer whereas it should be a long.

## Questions

Is this a known bug in version 44?
Are long integers supported in more recent versions of ojAlgo?

Cheers

**Comment by apete** [comment]
1) You should always try with the latest version before "asking"
2) It's not clear to me exactly where you have those large numbers, but I can tell you:
    a) Believe I have seen StackOverflowException on some very large MIP models (that I assume ojAlgo won't be able to solve anyway)
    b) In the MIP solver the lower/upper bounds of the integer variables are stored as `int[]`. Is this your problem? When you model your problem (using ExpressionBasedModel) you should NOT used very large numbers to indicate "no limit". If there is no limit then don't set a limit!

**Reply by ahonorat** [reply]
Hello @apete, thanks for your reply.

1. I used ojAlgo version 44 because it is a static dependency of the software I'm using. I'm asking this here *because* I cannot change the version of this dependency myself. Unfortunately, even the maintainer cannot do it before some months most probably (complex buggy procedure in our product).

2. a) I have large integers in weights of the objectives as well as in coefficients of variables in the linear expressions. Lower limits of linear expressions are also large integers. However in general, I guess my variables should not reach the integer limit when the model is optimized.
    b) Sure I did not set upper limit on variables to minimize in the end (it was just a try). Should I deduce that all other numbers (coefficients, expression bounds) also are stored with int32?

**Comment by apete** [comment]
In `ExpressionBasedModel` everything is stored as `BigDecimal` which is "unlimited".

The solvers typically work with `double`, and when the solvers are instantiated the problem parameters are scaled to improve numerical accuracy. However, the scaling functionality is disabled when exceptionally large parameters are encountered. In those cases expressions are instead used "as is".

The MIP solver internally keeps track of lower/upper integer bounds using `int[]`.

Not sure what your problem is, but if possible I'd try to scale the parameters "manually" before setting them in `ExpressionBasedModel`.

Not sure upgrading to the latest version would solve this particular problem, but the various solvers have been much improved with recent releases. I'd make an effort to try and upgrade to the latest version.

**Reply by ahonorat** [reply]
Ok, well I've tried multiple formulations but my coefficients/limits are quite unbalanced and I always end up with one being quite large.
But according to your answer maybe I should avoid large numbers at least for the lower bounds.
If some day I manage to update the version, I'll update this discussion if it solves the problem.

Thank you
